pcap-hurd.c

```diff
@@ -34,10 +34,10 @@ static struct bpf_insn filter[] = {
 	{ BPF_RET | BPF_K, 0, 0, MAXIMUM_SNAPLEN },
 };
 
-#define FILTER_COUNT (sizeof(filter) / sizeof(short))
+#define FILTER_COUNT (sizeof(filter) / (sizeof(short)))
```

Comment:
What was the warning with this line?

Reply:
gcc complains that the division is possibly incorrect because it assumes that we wanted to count the number of elements, and we are not dividing the size of the array (in bytes) by the size of an element (in bytes) to get the number of elements. And indeed, bpf programs are counted in number of shorts.

Reply:
This one is from Clang 11.0 only:
```
./pcap-hurd.c:196:35: warning: expression does not compute the number of elements in this array; element type is 'struct bpf_insn', not 'short' [-Wsizeof-array-div]
                          (filter_array_t_filter, FILTER_COUNT);
                                                  ^~~~~~~~~~~~
./pcap-hurd.c:37:38: note: expanded from macro 'FILTER_COUNT'
#define FILTER_COUNT (sizeof(filter) / sizeof(short))
                             ~~~~~~  ^
./pcap-hurd.c:32:24: note: array 'filter' declared here
static struct bpf_insn filter[] = {
                       ^
./pcap-hurd.c:196:35: note: place parentheses around the 'sizeof(short)' expression to silence this warning
./pcap-hurd.c:37:38: note: expanded from macro 'FILTER_COUNT'
#define FILTER_COUNT (sizeof(filter) / sizeof(short))
                                     ^
```
This looks a genuine bug to me, should have been `sizeof(filter[0])`.

The other three are visible with both Clang 11.0 and GCC 10.2.

Reply:
If the size uses `sizeof(short)` as the unit, it would be useful to have a comment about it.

Reply:
No, FILTER_COUNT really has to be the size in terms of shorts, that's how bpf is interfaced in device_set_filter, if you divide by `sizeof(filter[0])` things will get completely wrong.

It can look like an unfortunate choice, but it happens to be how it is, just like other set_status etc. calls of the Mach kernel.

Reply:
The other size of device_set_filter is implemented here: https://git.savannah.gnu.org/cgit/hurd/hurd.git/tree/libbpf/bpf_impl.c#n589 with CSPF_BYTES defined here: https://git.savannah.gnu.org/cgit/hurd/hurd.git/tree/libbpf/bpf_impl.h#n93, filter_t being a short.

Reply:
> If the size uses `sizeof(short)` as the unit, it would be useful to have a comment about it.

done so

Reply:
You might want to expand the comment to indicate that, while the parentheses around `sizeof(short)` are not required by the syntax of C, they're there to suppress a warning from some compilers that know about the "divide the size of an array by the size of an array element in order to find the number of elements of the array" idiom in C, and see that the divisor *isn't* the size of the array and warn that this *isn't* computing the number of elements in the array, and use the extra parentheses as an indication that the programmer is saying "trust me, this is correct even though the divisor isn't the size of an array element", so that we can forestall people submitting "helpful" patches/pull requests that remove the "unnecessary" parentheses.